org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java

```diff
@@ -0,0 +1,582 @@
+					this.e= new Elem(qn.getQualifier(), flag);
+			} else if (node instanceof FieldAccess) {
+				FieldAccess fa= (FieldAccess) node;
+				SimpleName sn= fa.getName();
+				this.memberKey= sn.resolveBinding() != null ? sn.resolveBinding().getKey() : null;
+				ASTNode expr= getOriginalExpression(fa.getExpression());
+				if (expr instanceof MethodInvocation) {
+					if (flag) {
+						this.e= new Elem((MethodInvocation) expr);
+					} else {
+						this.memberKey= null;
+					}
+				} else
+					this.e= new Elem(expr, flag);
+			} else if (node instanceof MethodInvocation) {
+				MethodInvocation mi= (MethodInvocation) node;
+				if (flag) {
+					this.e= new Elem(mi);
+				} else {
+					this.memberKey= null;
+				}
+			}
+		}
+
+		public Elem(MethodInvocation expr) { // use String to represent the instances of MethodInvocation
+			super();
+			this.memberKey= expr.toString();
+		}
+
+		public Elem(ASTNode node, boolean flag, Elem e) {
+			if (node instanceof MethodInvocation && flag) {
+				MethodInvocation mi= (MethodInvocation) node;
+				this.e= new Elem(mi.getExpression(), flag);
+				this.memberKey= e;
+			}
+		}
+
+		@Override
+		public int hashCode() {
+			final int prime= 31;
+			int result= 1;
+			result= prime * result + getEnclosingInstance().hashCode();
+			result= prime * result + this.toString().hashCode();
+			return result;
+		}
+
+		@Override
+		public String toString() {
+			String str1= this.e == null ? "" : this.e.toString(); //$NON-NLS-1$
+			String str2= ""; //$NON-NLS-1$
+			if (this.memberKey instanceof String) {
+				str2= (String) this.memberKey;
+			} else if (this.memberKey instanceof Elem) {
+				str2= ((Elem) this.memberKey).toString();
+			}
+			if (!str1.equals("")) //$NON-NLS-1$
+				return str1 + str2;
+			else
+				return str2;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj) {
+				return true;
+			}
+			if (!(obj instanceof Elem)) {
+				return false;
+			}
+			Elem other= (Elem) obj;
+			if (!getEnclosingInstance().equals(other.getEnclosingInstance())) {
+				return false;
+			}
+			String str1= this.toString();
+			String str2= obj.toString();
+			return str1.equals(str2);
+		}
+
+		private ChangedValueChecker getEnclosingInstance() {
+			return ChangedValueChecker.this;
+		}
+	}
+
+	class PathVisitor extends ASTVisitor {
+		ArrayList<ASTNode> nodes;
+
+		ArrayList<IASTFragment> candidateList;
+
+		HashSet<Position> posSet;
+
+		int startOffset;
+
+		int endOffset;
+
+		int type;
+
+		ASTNode selectedExpression;
+
+		public PathVisitor(int startOffset, int endOffset, ASTNode node, ArrayList<IASTFragment> candidateList) {
+			nodes= new ArrayList<>();
+			posSet= new HashSet<>();
+			this.startOffset= startOffset;
+			this.endOffset= endOffset;
+			this.selectedExpression= node;
+			this.candidateList= candidateList;
+			type= 2;
+			extend2EndOfLoop(selectedExpression);
+		}
+
+		private void extend2EndOfLoop(ASTNode node) {
+			ASTNode temp= node;
```

Comment:
Do you mean to go to end of outermost loop or did you mean to stop at innermost loop?  The code below will find the end of outermost loop and if you add a statement to innermost, it cannot be used in outer loops.

Reply:
This method is used for searching the end of the outermost loop in Sub-AST, which has the same parent node as the first expression. For example, 
```java
while (x < 100) {
	int y = calculateCount(x);
	while (x < 10) {
		for (int i = 1; i < 10; ++i) {
			for (int j = 1; j < 10; ++j) {
				int m = calculateCount(x);
				int z = x;
			}
		}
	}
	x++;
}
```
The path between `int y = calculateCount(x);` and `int m = calculateCount(x);` is shown as follows:
```java
int y = calculateCount(x);
while (x < 10) {
for (int i = 1; i < 10; ++i) {
	for (int j = 1; j < 10; ++j) {
		int m = calculateCount(x);
		int z = x;
	}
}
```

Reply:
Ok